File: 프로그래머스/3/258709. 주사위 고르기/주사위 고르기.java
import java.util.*;

class Solution {
    static List<List<Integer>> combs = new ArrayList<>();
    
    public int[] solution(int[][] dice) {
        int[] answer;
        
        int n = dice.length; // 주사위의 개수
        int maxWin = -1;
        List<Integer> bestComb = new ArrayList<>();
        
        boolean[] visited = new boolean[n];
        combinations(dice, visited, 0, n, n/2);
        
        for (List<Integer> comb: combs) {
            List<int[]> A = new ArrayList<>();
            for (Integer a: comb) A.add(dice[a]);
            
            List<int[]> B = new ArrayList<>();
            for (int i = 0; i < n; i++) {
                if (!comb.contains(i)) B.add(dice[i]);
            }
            
            int[] ASums = getSums(A);
            int[] BSums = getSums(B);
            
            int winCases = 0;
            for (int a: ASums) {
                winCases += findLowerCnt(BSums, a);
            }
            
            if (winCases > maxWin) {
                maxWin = winCases;
                bestComb = comb;
            }
        }
        answer = bestComb.stream().mapToInt(i->i+1).toArray();
        return answer;
    }
    
    public static void combinations(int[][] arr, boolean[] visited, int start, int n, int r) {
        if (r == 0) {
            List<Integer> comb = new ArrayList<>();
            for (int i = 0; i < n; i++) {
                if (visited[i]) comb.add(i);
            }
            combs.add(comb);
            return;
        }
        
        for (int i = start; i < n; i++) {
            visited[i] = true;
            combinations(arr, visited, i+1, n, r-1);
            visited[i] = false;
        }
    }
    
    public static int[] getSums(List<int[]> dices) {
        List<Integer> sums = new ArrayList<>();
        dfs(dices, 0, 0, sums);
        Collections.sort(sums);
        return sums.stream().mapToInt(i->i).toArray();
    }
    
    public static void dfs(List<int[]> dices, int idx, int currSum, List<Integer> sums) {
        if (idx == dices.size()) {
            sums.add(currSum);
            return;
        }
        for (int face: dices.get(idx)) {
            dfs(dices, idx+1, currSum + face, sums);
        }
    }
    
    public static int findLowerCnt(int[] arr, int target) {
        int left = 0, right = arr.length;
        
        while (left < right) {
            int mid = (left + right) / 2;
            if (arr[mid] < target) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        return left;
    }
}

File: 프로그래머스/3/258709. 주사위 고르기/주사위 고르기.py
from itertools import combinations

def get_sums(dices):
    # 주사위 여러 개가 주어졌을 때, 각 주사위에서 하나씩 눈을 골라 나올 수 있는 '모든 합'을 구한다.
    sums = []
    
    def dfs(idx, curr_sum):
        if idx == len(dices):
            sums.append(curr_sum)
            return
        for face in dices[idx]:
            dfs(idx+1, curr_sum + face)
    
    dfs(0, 0)
    sums.sort()
    return sums

def find_lower_cnt(arr, target):
    # arr에서 target보다 작은 값의 개수를 반환
    # arr에 target이 추가된 후 arr를 정렬했을 때 target의 인덱스가 된다.
    left, right = 0, len(arr)
    
    while left < right:
        mid = (left + right) // 2
        if arr[mid] < target:
            left = mid + 1
        else:
            right = mid
    return left

def solution(dice):
    answer = []
    
    n = len(dice) # 주사위의 개수
    max_win = -1 # 가장 높은 승리 횟수
    
    for comb in combinations(range(n), n//2):
        # comb는 A가 가져가는 주사위 번호
        A = [dice[i] for i in comb]
        B = [dice[i] for i in range(n) if i not in comb]
        
        A_sums = get_sums(A)
        B_sums = get_sums(B)
        
        win_cases = 0
        for a in A_sums:
            # B의 합 중에서 a보다 작은 값 개수 세기
            win_cases += find_lower_cnt(B_sums, a)
        
        if win_cases > max_win: # 승리할 확률이 가장 높은 주사위 조합 갱신
            max_win = win_cases
            answer = comb
        
    return [i+1 for i in answer]